Projects/UOContent/Mobiles/AI/BaseAI/AIGroupMovement.cs

```diff
@@ -0,0 +1,222 @@
+ * You should have received a copy of the GNU General Public License     *
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.  *
+ ************************************************************************/
+
+using System;
+using System.Collections.Generic;
+
+namespace Server.Mobiles;
+
+public abstract partial class BaseAI
+{
+    private static readonly Dictionary<BaseCreature, Point3D> _reservedPositions = new();
+    private static long _lastGroupUpdateTime = 0;
+
+    private static void CleanupReservedPositions()
+    {
+        var toRemove = new List<BaseCreature>();
+
+        foreach (var kvp in _reservedPositions)
+        {
+            if (kvp.Key == null || kvp.Key.Deleted || kvp.Key.GetDistanceToSqrt(kvp.Value) < 1)
+            {
+                toRemove.Add(kvp.Key);
+            }
+        }
+        foreach (var creature in toRemove)
+        {
+            _reservedPositions.Remove(creature);
+        }
+    }
+
+    private bool UseGroupMovement(Mobile target) =>
+        m_Mobile.Combatant == target
+        && !m_Mobile.Controlled
+        && GetNearbyAllies(target).Count > 0;
+
+    public static bool MoveToWithGroup(BaseAI ai, Mobile target, bool run, int range)
+    {
+        if (Core.TickCount - _lastGroupUpdateTime > 1000)
+        {
+            CleanupReservedPositions();
+            _lastGroupUpdateTime = Core.TickCount;
+        }
+
+        var m_Mobile = ai.m_Mobile;
+        var allies = ai.GetNearbyAllies(target);
+        var optimalPosition = ai.CalculateOptimalPosition(target, allies, range);
+
+        if (optimalPosition != Point3D.Zero)
+        {
+            _reservedPositions[m_Mobile] = optimalPosition;
+
+            var direction = m_Mobile.GetDirectionTo(optimalPosition);
+
+            if (Utility.Random(100) < 30)
+            {
+                direction = GetAdjustedDirection(direction);
+            }
+
+            return ai.DoMove(direction, true);
+        }
+
+        return ai.MoveToWithCollisionAvoidance(target, run, range);
+    }
+
+    private List<BaseCreature> GetNearbyAllies(Mobile target)
+    {
+        var mobiles = new List<Mobile>();
+
+        foreach (var m in m_Mobile.GetMobilesInRange(8))
+        {
+            mobiles.Add(m);
+        }
+
+        if (mobiles.Count <= 1)
+        {
+            return [];
+        }
+
+        var allies = new List<BaseCreature>();
+
+        foreach (var mobile in mobiles)
+        {
+            if (mobile is BaseCreature bc && bc != m_Mobile
+                                          && bc.Combatant == target && !bc.Controlled && bc.Team == m_Mobile.Team)
+            {
+                allies.Add(bc);
+            }
+        }
+
+        return allies;
+    }
+
+    private Point3D CalculateOptimalPosition(Mobile target, List<BaseCreature> allies, int range)
+    {
+        var targetLoc = target.Location;
+        var positions = new List<Point3D>();
```

Comment:
The allocation can be eliminated entirely by using a sliding window (tracking bestPosition/bestScore as you go through the initial for loops)